attachments/utils.py

```diff
@@ -35,25 +35,36 @@ def __init__(
         attachment=None,
         kind=None,
         force_desiredness=None,
+        optionality_group=None,
     ):
         self.attachment = attachment
         self.attached_object = attached_object
         self.kind = kind
         self.force_desiredness = force_desiredness
+        self.optionality_group = optionality_group
+        if self.optionality_group:
+            self.optionality_group.members.add(self)
 
-    def match(self, exclude):
+    def match(self, exclude=[]):
         """
-        Tries to fill this slot with an existing attachment that is not
-        in the exclusion set of already matched attachments. Returns True
-        or False depending on if the slot was succesfully matched.
+        Tries to find a matching attachment for this slot. If it finds one,
+        it returns the attachment, otherwise it returns False.
         """
         for instance in self.get_instances_for_slot():
             if instance not in exclude:
-                self.attachment = instance
-                self.kind = get_kind_from_str(instance.kind)
-                return True
+                return instance
         return False
 
+    def match_and_set(self, exclude):
+        """
+        Uses self.match() to find a matching attachment. If it finds one, it
+        sets self.attachment and self.kind
+        """
+        matched_attachment = self.match(exclude=exclude)
+        if matched_attachment:
+            self.attachment = matched_attachment
+            self.kind = get_kind_from_str(matched_attachment.kind)
```

Comment:
Match_and_set should actually still return True, or the instance, or anything truthy if it succeeds. Otherwise `stepper.attachment_slots()` will not work.

Although in that specific method we do actually want to modify internal slot state.